packages/@aws-cdk/aws-ec2/lib/vpc-endpoint-service.ts

```diff
@@ -50,6 +50,19 @@ export class VpcEndpointService extends Resource implements IVpcEndpointService
    */
   public readonly whitelistedPrincipals: ArnPrincipal[];
 
+  /**
+   * The id of the VPC Endpoint Service, like vpce-svc-xxxxxxxxxxxxxxxx.
+   * @attribute
+   */
+  public readonly vpcEndpointServiceId: string;
+
+  /**
+   * The service name of the VPC Endpoint Service that clients use to connect to,
+   * like com.amazonaws.vpce.<region>.vpce-svc-xxxxxxxxxxxxxxxx
+   * @attribute
+   */
+  public readonly serviceName: string;
```

Comment:
`vpcEndpointServiceName` to be consistent with other names.

Reply:
Ah, that one already exists. In that case, if `serviceName` means two distinct things, is this really the best name?

Reply:
This is a long story. I wanted to use vpcEndpointServiceName for that purpose, but repeatedly failed to make the code compile. 

The problem seems to stem from vpcEndpointServiceName being an optional property in the IVpcEndpointService. If I try and set vpcEndpointServiceName in the VpcEndpointService class, I get this error:

> error TS0: @aws-cdk/aws-ec2.VpcEndpointService#vpcEndpointServiceName changes optionality of property when implementing @aws-cdk/aws-ec2.IVpcEndpointService

I am not sure exactly how to fix this. I tried making the property non-optional, but got this error. This seems to suggest that I have to add additional code beyond the scope of what I'm trying to do:

> error: [awslint:from-method:@aws-cdk/aws-ec2.VpcEndpointService] resource should have at least one "fromXxx" static method or "fromXxxAttributes" 

I tried removing the property, which lets the code compile, but then IVpcEndpointService would be an empty extension of IResource.

Any recommendations on what to do?

Reply:
> I am not sure exactly how to fix this. I tried making the property non-optional, but got this error.

Yes. You can still declare it as optional, and still always give it a value though, which would solve that issue.

```ts
  public readonly vpcEndpointServiceName: string | undefined = 'hello';
```

Is perfectly legal. You're telling consumers that the value *could* be missing (even though in practice it never is). It's not really a contradiction, just maybe a little inconvenient.

But I don't like the way `vpcEndpointServiceName` looks today, at all.

Right now:

```
class VpcEndpointServiceProps {
  readonly vpcEndpointServiceName?: string;
}
```

Even though it is never used anywhere! Nor does the [CloudFormation Resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-vpcendpointservice.html#aws-resource-ec2-vpcendpointservice-return-values) even seem to accept something looking like a name. So what's the point of this property?

And then there is:

```ts
export interface IVpcEndpointService extends IResource {
  /**
   * Name of the Vpc Endpoint Service
   * @experimental
   */
  readonly vpcEndpointServiceName?: string;
}
```

Even though this property is never set. So yes, it's very optional.

----

Would you be so kind to fix this as part of this PR? I think we should do the following:

* Add `@deprecate This property is not used.` to the Props field.
* Make the `IVpcEndpointService` field non-optional, and then use it for your name calculation.
* Add the `com.amazonaws.vpce` prefix to `@aws-cdk/region-info` (and the Chinese variant) and look it up from there. You should be able to grep around the codebase for examples on how to do that, if you're having trouble I can give you some pointers.

I think that would bring it in line with what we expect from the construct.

Sorry for blowing up this PR on you :)

Reply:
I like everything you've said, I'm happy to expand the scope of the PR -- I'd rather it be done right.

Reply:
How can I avoid the error message I was getting before, where changing the property optionality resulted in this message:

>  error: [awslint:from-method:@aws-cdk/aws-ec2.VpcEndpointService] resource should have at least one "fromXxx" static method or "fromXxxAttributes"

Reply:
Ah, I found it, in package.json. Will add a lint rule to exclude VpcEndpointService from this check.